src/docfx/validation/ContentValidator.cs

```diff
@@ -43,6 +44,89 @@ public void ValidateManifest(FilePath filePath, PublishItem publishItem)
             }
         }
 
+        public void ValidateTocDeprecated(FilePath filePath)
+        {
+            if (TryGetValidationDocumentType(ContentType.TableOfContents, string.Empty, false, out var documentType))
+            {
+                var validationContext = new ValidationContext { DocumentType = documentType };
+                var tocItem = new TocItem()
+                {
+                    FilePath = filePath.Path.Value,
+                    SourceInfo = new SourceInfo(filePath, 0, 0),
+                };
+                Write(_validator.ValidateToc(tocItem, validationContext).GetAwaiter().GetResult());
+            }
+        }
+
+        public void ValidateTocMissing(Document document, bool hasReferencedTocs)
+        {
+            if (TryGetValidationDocumentType(document.ContentType, document.Mime.Value, false, out var documentType))
+            {
+                var validationContext = new ValidationContext { DocumentType = documentType };
+                var tocItem = new TocItem()
+                {
+                    FilePath = document.FilePath.Path.Value,
+                    HasReferencedTocs = hasReferencedTocs,
+                    SourceInfo = new SourceInfo(document.FilePath, 0, 0),
+                };
+                Write(_validator.ValidateToc(tocItem, validationContext).GetAwaiter().GetResult());
+            }
+        }
+
+        public void ValidateBreadcrumbLinkExternal(TableOfContentsNode[] nodes, FilePath filePath)
+        {
+            if (TryGetValidationDocumentType(ContentType.TableOfContents, string.Empty, false, out var documentType))
+            {
+                var validationContext = new ValidationContext { DocumentType = documentType };
+                foreach (var node in nodes)
+                {
+                    var tocItem = new TocItem()
+                    {
+                        FilePath = node.Href.Value,
+                        IsHrefExternal = UrlUtility.GetLinkType(node.Href) == LinkType.External,
+                        SourceInfo = new SourceInfo(filePath, 0, 0),
+                    };
+                    Write(_validator.ValidateToc(tocItem, validationContext).GetAwaiter().GetResult());
```

Comment:
If I understand correctly , the `ValidateToc` call here looks through all the TOC validators [here](https://dev.azure.com/ceapex/Engineering/_git/Docs.Validation?path=%2FValidation%2FValidator.cs&version=GBdevelop&line=65&lineEnd=65&lineStartColumn=39&lineEndColumn=67&lineStyle=plain), and so even if this method is called `ValidateBreadcrumbLinkExternal`, it'll still execute all 4 TOC validators against each TOC node. If so, there are a couple of problems I see with this design:

1. It is very easy to introduce bugs. The current tests all checks against a single rule, if there are multiple rules turned on at the same time, e.g., if `TocMissing` and `TocBreadcrumbLinkExternal` are both turned on, will there be a lot of `TocMissing` errors for breadcrumb since the `ValidateTocMissing` validator is using the [HasReferencedTocs](https://dev.azure.com/ceapex/Engineering/_git/Docs.Validation?path=%2FValidation%2FValidateToc%2FValidateTocMissing.cs&version=GBdevelop&line=20&lineEnd=20&lineStartColumn=26&lineEndColumn=43&lineStyle=plain) property for detection, and each breadcrumb TOC node does not set `HasReferencedTocs`?

2. The performance is sub-optimal.

Reply:
I have changed the interface, now it's only run one rule for each toc model type.